src/quick-lint-js/fe/parse-statement.cpp

```diff
@@ -1880,36 +1880,43 @@ void Parser::parse_and_visit_statement_block_no_scope(Parse_Visitor_Base &v) {
 void Parser::parse_and_visit_statement_block_no_scope(
     Parse_Visitor_Base &v, Parse_Statement_Options statement_options) {
   QLJS_ASSERT(this->peek().type == Token_Type::left_curly);
-  Source_Code_Span left_curly_span = this->peek().span();
+  Token left_curly = this->peek();
   this->skip();
-  this->parse_and_visit_statement_block_after_left_curly(v, left_curly_span,
+  this->parse_and_visit_statement_block_after_left_curly(v, left_curly,
                                                          statement_options);
 }
 
 void Parser::parse_and_visit_statement_block_after_left_curly(
-    Parse_Visitor_Base &v, Source_Code_Span left_curly_span) {
+    Parse_Visitor_Base &v, Token left_curly) {
   this->parse_and_visit_statement_block_after_left_curly(
-      v, left_curly_span,
+      v, left_curly,
       Parse_Statement_Options{
           .possibly_followed_by_another_statement = true,
       });
 }
 
 void Parser::parse_and_visit_statement_block_after_left_curly(
-    Parse_Visitor_Base &v, Source_Code_Span left_curly_span,
+    Parse_Visitor_Base &v, Token left_curly,
     Parse_Statement_Options statement_options) {
+  QLJS_ASSERT(left_curly.type == Token_Type::left_curly);
   for (;;) {
     bool parsed_statement =
         this->parse_and_visit_statement(v, statement_options);
     if (!parsed_statement) {
       switch (this->peek().type) {
       case Token_Type::right_curly:
+        if (this->peek().indent_level != left_curly.indent_level) {
+          this->diag_reporter_->report(Diag_Mismatched_Curly_Indentation{
+              .left_curly = left_curly.span(),
+              .right_curly = this->peek().span(),
+          });
```

Comment:
I removed this diagnostic completely in this new set of patches. I replaced it with `Diag_Unclosed_Code_Block_V2`, which I think matches your intention in #282 much more closely.

It will not report any warnings for the snippet of code above. But it will report an error for the snippet of code in #282 as well as others obviously. See the tests for more examples.

Note, this new diagnostic does not catch something like this, where more code trails an expected but missing closing curly:
```javascript
function f() {
  z;
  if (false) {
    z;
    if (1) {
      z;
    }

  if (true) {
  }
}
```
It will report that the curly brace after `if (false)` is missing its matching curly. However, it will suggest the matching rightly curly go at the end of the file instead of before `if (true)`.